examples/step-57/step-57.cc

```diff
@@ -0,0 +1,948 @@
+    const SparseMatrix<double>      &pressure_mass_matrix;
+    const PreconditionerMp          &mp_preconditioner;
+  };
+
+  template <class PreconditionerMp>
+  BlockSchurPreconditioner<PreconditionerMp>::
+  BlockSchurPreconditioner (double                           gamma,
+                            double                           viscosity,
+                            const BlockSparseMatrix<double>  &S,
+                            const SparseMatrix<double>       &P,
+                            const PreconditionerMp           &Mppreconditioner)
+    :
+    gamma                (gamma),
+    viscosity            (viscosity),
+    stokes_matrix        (S),
+    pressure_mass_matrix (P),
+    mp_preconditioner    (Mppreconditioner)
+  {}
+
+  template <class PreconditionerMp>
+  void
+  BlockSchurPreconditioner<PreconditionerMp>::
+  vmult (BlockVector<double>       &dst,
+         const BlockVector<double> &src) const
+  {
+    Vector<double> utmp(src.block(0));
+
+    {
+      SolverControl solver_control(1000, 1e-6 * src.block(1).l2_norm());
+      SolverCG<>    cg (solver_control);
+
+      dst.block(1) = 0.0;
+      cg.solve(pressure_mass_matrix,
+               dst.block(1), src.block(1),
+               mp_preconditioner);
+      dst.block(1) *= -(viscosity+gamma);
+    }
+
+    {
+      stokes_matrix.block(0,1).vmult(utmp, dst.block(1));
+      utmp*=-1.0;
+      utmp+=src.block(0);
+    }
+
+    SparseDirectUMFPACK  A_direct;
+    A_direct.initialize(stokes_matrix.block(0,0));
+    A_direct.vmult (dst.block(0), utmp);
+  }
+
+  // @sect3{Navier_Stokes_Newton class implementation}
+  // @sect4{Navier_Stokes_Newton::Navier_Stokes_Newton}
+  // The constructor of this class looks very similar to the one in step-22. The only difference is the
+  // viscosity and the AL coefficient gamma. In test case, if the viscosity is relatively large (1/400), the initial guess
+  // can be obtained by solving a corresponding Stokes problem. However, auxiliary NSE is necessary if viscosity is relatively
+  // small (1/10000).
+
+  template <int dim>
+  Navier_Stokes_Newton<dim>::Navier_Stokes_Newton(const unsigned int degree)
+    :
+
+    viscosity(1.0/400.0),
+    gamma(1.0),
+    degree(degree),
+    triangulation(Triangulation<dim>::maximum_smoothing),
+    fe(FE_Q<dim>(degree+1), dim,
+       FE_Q<dim>(degree),   1),
+    dof_handler(triangulation)
+  {}
+
+
+  template <int dim>
+  Navier_Stokes_Newton<dim>::~Navier_Stokes_Newton()
+  {
+    dof_handler.clear();
+  }
+
+  // @sect4{Navier_Stokes_Newton::setup_dof}
+  // This function provide the information of DoF and initialize the constraints on current mesh.
+
+  template <int dim>
+  void Navier_Stokes_Newton<dim>::setup_dof()
+  {
+    system_matrix.clear();
+    pressure_mass_matrix.clear();
+
+    // The first step is to associate DoFs with a given mesh. Here it is completed as in step-22
+    dof_handler.distribute_dofs (fe);
+    DoFRenumbering::Cuthill_McKee (dof_handler);
```

Comment:
Why do you use Cuthill_McKee? You use a direct solver in the velocity, which is not affected by the ordering. And the ILU for a mass matrix should converge quickly even standard ordering. And mat-vecs will probably be slightly faster with our default numbering - to the extent cache effects for vector access matters when streaming the matrix entries is the big factor. (But I'd be happy to be proven wrong by actual computational results.)